onnxruntime/core/optimizer/reshape_fusion.cc

```diff
@@ -120,35 +120,72 @@ bool ReshapeFusion::Fuse_Subgraph1(Node& reshape, Graph& graph, const logging::L
     return false;
   }
 
-  // Loop through the inputs of concat node to calculate the shape_value for a potential reshape fusion. 
+  // Loop through the inputs of concat node to calculate the shape_value for a potential reshape fusion.
   std::vector<int64_t> shape_value;
   shape_value.reserve(concat_input_count);
   for (int i = 0; i < concat_input_count; ++i) {
     // First check if the i-th argument is a constant initializer.
     if (optimizer_utils::AppendTensorFromInitializer(graph, *(concat.InputDefs()[i]), shape_value, true)) {
       continue;
     }
+
     // Try to find path [Root] --> Shape --> Gather(indices=i) --> Unsqueeze (axes=0) --> Concat [input i]
     bool matched = ReshapeFusion::Fuse_Subgraph2(graph, root_input, concat, i, shape_value, logger);
-    const Node* p_cur_node = graph_utils::GetInputNode(concat, i);
     if (matched) {
       shape_value.push_back(0);
-    } else if (p_cur_node != nullptr) {
-      // This node could lead to a potential subgraph pattern fusion. Mark the shape value to -1.
-      shape_value.push_back(-1);
-    } else {
+      // We have matched the pattern for this input into Concat
+      // Proceed to the next input
+      continue;
+    }
+
+    // If we haven't been able to match the pattern, try and see if this is a candidate for subgraph pattern fusion.
+    // For this input to be a candidate, the number of elements in the input tensor to Concat has to be 1
+    // We use shape info (if made available via shape inference) for this.
+    const NodeArg* concat_input_node_arg = concat.InputDefs()[i];
+    if (!concat_input_node_arg) {
+      // We need NodeArg to be able to lookup any shape info
+      // Can't proceed with fusion
       return false;
     }
-  }  
 
```

Comment:
I don't think this can be hit. We create a NodeArg for all node inputs and outputs when loading the graph (Graph::AddNode calls Graph::CreateNodeArgs), and as the concat input is variadic there won't be missing optional inputs in the middle of them.

Reply:
Removed check